spec/spec_tests/uri_options_spec.rb

```diff
@@ -0,0 +1,123 @@
+
+      before(:all) do
+
+        module Mongo
+          class Address
+
+            private
+
+            alias :original_create_resolver :create_resolver
+            def create_resolver(timeout, ssl_options)
+              family = (host == 'localhost') ? ::Socket::AF_INET : ::Socket::AF_UNSPEC
+              info = ::Socket.getaddrinfo(host, nil, family, ::Socket::SOCK_STREAM)
+              FAMILY_MAP[info.first[4]].new(info[3], port, host)
+            end
+          end
+
+          class Server
+
+            # The constructor keeps the same API, but does not instantiate a
+            # monitor and run it.
+            alias :original_initialize :initialize
+            def initialize(address, cluster, monitoring, event_listeners, options = {})
+              @address = address
+              @cluster = cluster
+              @monitoring = monitoring
+              @options = options.freeze
+              @monitor = Monitor.new(address, event_listeners, Monitoring.new, options)
+            end
+
+            # Disconnect simply needs to return true since we have no monitor and
+            # no connection.
+            alias :original_disconnect! :disconnect!
+            def disconnect!; true; end
+          end
+        end
+      end
+
+      after(:all) do
+
+        module Mongo
+          # Return the implementations to their originals for the other
+          # tests in the suite.
+          class Address
+            alias :create_resolver :original_create_resolver
+            remove_method(:original_create_resolver)
+          end
+
+          class Server
+            alias :initialize :original_initialize
+            remove_method(:original_initialize)
+
+            alias :disconnect! :original_disconnect!
+            remove_method(:original_disconnect!)
+          end
+        end
+      end
+
+      spec.tests.each do |test|
+
+        # Skip optional tests for options the driver doesn't support.
+        if test.description.start_with?('tlsAllowInvalidHostnames') || test.description.start_with?('tlsInsecure')
+          next
+        end
+
+        context "#{test.description}" do
+
+          context 'when the uri is invalid', unless: test.valid? do
+
+            it 'raises an error' do
+              expect{
+                test.uri
+              }.to raise_exception(Mongo::Error::InvalidURI)
+            end
+          end
+
+          context 'when the uri should warn', if: test.warn? do
+
+            before do
+              expect(Mongo::Logger.logger).to receive(:warn)
+            end
+
+            it 'warns' do
+              expect(test.client).to be_a(Mongo::Client)
+            end
+          end
+
+          context 'when the uri is valid', if: test.valid? && !test.warn? do
+
+            before do
+              expect(Mongo::Logger.logger).not_to receive(:warn)
+            end
+
+            it 'does not raise an exception or warning' do
+              expect(test.uri).to be_a(Mongo::URI)
```

Comment:
Why is this test not performed when `test.valid? && test.warn?` ?

Reply:
Fixed